File: app/src/main/java/com/demo/util/SimpleMethodInterceptor.java
package com.demo.util;

import net.sf.cglib.proxy.MethodInterceptor;
import net.sf.cglib.proxy.MethodProxy;

import java.lang.reflect.Method;
import java.util.Arrays;

public class SimpleMethodInterceptor implements MethodInterceptor {
    /**
     * @param obj    "this", the enhanced object (i.e. a child class instance)
     * @param method intercepted Method
     * @param args   argument array; primitive types are wrapped
     * @param proxy  used to invoke super (non-intercepted method); may be called
     *               as many times as needed
     */
    @Override
    public Object intercept(Object obj, Method method, Object[] args, MethodProxy proxy) throws Throwable {
        // "obj" is an instance of the child class,
        // proxy.invokeSuper(...) will finally invoke corresponding method in parent class (i.e. "IntegerCalculator" class)
        Object result = proxy.invokeSuper(obj, args);

        System.out.println("params: " + Arrays.toString(args));
        System.out.println("result: " + result);
        System.out.println();

        return result;
    }
}

File: app/src/main/kotlin/com/demo/util/SimpleMethodInterceptor.kt
package com.demo.util

import net.sf.cglib.proxy.MethodInterceptor

class SimpleMethodInterceptor : MethodInterceptor {
    /**
     * @param obj    "this", the enhanced object (i.e. a child class instance)
     * @param method intercepted Method
     * @param args   argument array; primitive types are wrapped
     * @param proxy  used to invoke super (non-intercepted method); may be called
     * as many times as needed
     */
    @kotlin.Throws(Throwable::class)
    override fun intercept(
        obj: Any,
        method: java.lang.reflect.Method,
        args: Array<Any>,
        proxy: net.sf.cglib.proxy.MethodProxy
    ): Any {
        // "obj" is an instance of the child class,
        // proxy.invokeSuper(...) will finally invoke corresponding method in parent class (i.e. "IntegerAdded" class)
        val result = proxy.invokeSuper(obj, args)
        println("output from SimpleMethodInterceptor begins >>>")
        println("params: " + args.contentToString())
        println("result: $result")
        println("<<< output from SimpleMethodInterceptor ends")
        return result
    }
}